evaluator/evaluate_packages.go

```diff
@@ -137,233 +137,214 @@ func updatePackageNameDB(missing *models.PackageName) error {
-	id2index := make(map[int64]int, numUpdates)
-	i := 0
-	for nevra := range updates {
+	// allocate also space for removed packages
+	packages := make(map[string]namedPackage, numUpdates*2)
+	for nevra, pkgUpdate := range updates {
+		if !isValidNevra(nevra) {
+			continue
+		}
+		installed++
+		availableUpdates := pkgUpdate.GetAvailableUpdates()
+		if len(availableUpdates) > 0 {
+			updatable++
+		}
 		pkgMeta, ok := memoryPackageCache.GetByNevra(nevra)
+		// before we used nevra.EVRAString() function which shows only non zero epoch, keep it consistent
+		// maybe we need here something like: evra := strings.TrimPrefix(upData.GetEVRA(), "0:")
 		if ok {
-			packageIDs = append(packageIDs, pkgMeta.ID)
-			p := namedPackage{
-				NameID:    pkgMeta.NameID,
-				Name:      pkgMeta.Name,
-				PackageID: pkgMeta.ID,
-				EVRA:      pkgMeta.Evra,
-				WasStored: false,
+			installableID, applicableID := latestPackagesFromUpdatesList(availableUpdates)
+			packages[nevra] = namedPackage{
+				NameID:        pkgMeta.NameID,
+				Name:          pkgMeta.Name,
+				PackageID:     pkgMeta.ID,
+				EVRA:          pkgMeta.Evra,
+				Change:        Add,
+				InstallableID: installableID,
+				ApplicableID:  applicableID,
 			}
-			packages = append(packages, p)
-			id2index[pkgMeta.ID] = i
-			i++
 		}
 	}
-	rows, err := tx.Table("system_package2").
-		Select("package_id, installable_id, applicable_id").
+	utils.LogInfo("inventoryID", system.InventoryID, "packages", numUpdates)
+	return packages, installed, updatable
+}
+
+func loadSystemNEVRAsFromDB(tx *gorm.DB, system *models.SystemPlatform, packages map[string]namedPackage) error {
+	rows, err := tx.Table("system_package2 sp2").
+		Select("sp2.name_id, pn.name, sp2.package_id, p.evra, sp2.installable_id, sp2.applicable_id").
+		Joins("JOIN package p ON p.id = sp2.package_id").
+		Joins("JOIN package_name pn on pn.id = sp2.name_id").
 		Where("rh_account_id = ? AND system_id = ?", system.RhAccountID, system.ID).
-		Where("package_id in (?)", packageIDs).
 		Rows()
 	if err != nil {
-		return nil, err
+		return err
 	}
 	var columns namedPackage
+	numStored := 0
 	for rows.Next() {
 		err = tx.ScanRows(rows, &columns)
 		if err != nil {
-			return nil, err
+			return err
 		}
-		index := id2index[columns.PackageID]
-		packages[index].WasStored = true
-		packages[index].InstallableID = columns.InstallableID
-		packages[index].ApplicableID = columns.ApplicableID
+		numStored++
+		nevra := utils.NEVRAStringE(columns.Name, columns.EVRA, true)
+		if p, ok := packages[nevra]; ok {
+			if latestPkgsChanged(p, columns) {
+				p.Change = Update
+			} else {
+				p.Change = Keep
+			}
+		} else {
+			packages[nevra] = namedPackage{
+				NameID:        columns.NameID,
+				PackageID:     columns.PackageID,
+				EVRA:          columns.EVRA,
+				Change:        Remove,
+				ApplicableID:  columns.ApplicableID,
+				InstallableID: columns.InstallableID,
+			}
+		}
+	}
+	if err := rows.Close(); err != nil {
+		return err
 	}
```

Comment:
I think it would be better to call `defer rows.Close()` before the for loop to be sure that we released resources when ScanRows fails. To check for an error we might use `rows.Err()` after the for loop. `rows.Next()` should implicitly call `Close` once it iterates through all rows.
This seems to be the recommended way by golang docs https://go.dev/doc/database/querying
```go
defer rows.Close()
for rows.Next() {...}
if err := rows.Err(); err != nil { return err }
```

Reply:
For some reason `defer rows.Close()` failed with some tests in the past. I've rechecked it and now all tests pass so I've switched back to `defer ...`.